### PR TITLE
Notebook execution history

### PR DESCRIPTION
**Pull request by andrewfulton9**
## References

#9646

follows up on @echarles work [here](https://github.com/datalayer-externals/jupyterlab/tree/feat/notebookExecutionHistory)

## Code changes

- Adds keybindings "Alt ArrowUp" and "Alt ArrowDown" to notebook to iterate through kernel history in an activated cell
- Adds history code file to notebook for managing the kernel history iteration file
- Adds code to notebook widget file to add history functionality

## User-facing changes

This PR gives the user the ability to use Alt + up or down arrows to iterate through the kernel history. 

https://github.com/jupyterlab/jupyterlab/assets/13774419/e1532299-857d-4009-8b61-674152e06701

## Backwards-incompatible changes

None that I am aware

### CODE REVIEW

File: packages/notebook/src/widget.ts [hunk 1]
```diff
@@ -1033,6 +1036,11 @@ export namespace StaticNotebook {
      * The application language translator.
      */
     translator?: ITranslator;
+
+    /**
+     * The kernel history retrieval object
+     */
+    history?: INotebookHistory;
```
Comment (krassowski) [inline packages/notebook/src/widget.ts:1043]: Maybe worth being more specific as `inputHistory`? The other kinds that quickly come to mind are "editHistory" (undo stack), "checkpointHistory". I think this should be reflected in the name of attributes and interface.

File: packages/notebook/src/history.ts [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+    this._retrieveHistory(cell);
+  }
+
+  /**
+   * retrieve the history from the kernel
+   *
+   * @param cell - The string to use when filtering the data.
+   */
+  private async _retrieveHistory(cell?: Cell): Promise<void> {
+    return this._kernel?.requestHistory(Private.initialRequest).then(v => {
```
Comment (krassowski) [inline packages/notebook/src/history.ts:313]: Can be rewritten using await

File: packages/notebook/src/history.ts [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+   * @param activeCell - The currently selected Cell in the notebook.
+   *
+   * @returns A Promise for console command text or `undefined` if unavailable.
+   */
+  back(activeCell: Cell): Promise<string> {
+    this.checkSession(activeCell);
+    --this._cursor;
+    this._cursor = Math.max(0, this._cursor);
+    const content = this._filtered[this._cursor];
+    console.log('this._cursor: ', this._cursor);
```
Comment (krassowski) [inline packages/notebook/src/history.ts:180]:
```suggestion
```

File: packages/notebook-extension/schema/tracker.json [hunk 4]
```diff
@@ -366,6 +366,16 @@
     ]
   },
   "jupyter.lab.shortcuts": [
+    {
+      "command": "notebook:access-last-history",
+      "keys": ["Alt ArrowUp"],
```
Comment (krassowski) [inline packages/notebook-extension/schema/tracker.json:371]: Let's move these down to all other commands with `.jp-Notebook.jp-mod-editMode` selector?

File: packages/notebook/src/widget.ts [hunk 5]
```diff
@@ -1274,6 +1282,32 @@ export class Notebook extends StaticNotebook {
     this.addFooter();
   }
 
+  /**
+   * Access previous item in ipython kernel for current kernel
+   */
+  accessLastHistory(): void {
```
Comment (krassowski) [inline packages/notebook/src/widget.ts:1288]: I am worried about this being added on `Notebook` class - it extends the public API by two non-optional methods. What if instead the implementation was moved to respective `NotebookActions` functions (which currently just call these methods) and `history` getter were made public as an optional member on `Notebook`? This way is also easier to preserver backward compatibility. In this scenario we would review `INotebookHistory` to make it as narrow as possible.

File: packages/notebook/src/widget.ts [hunk 6]
```diff
@@ -1274,6 +1282,32 @@ export class Notebook extends StaticNotebook {
...
+    const activeCell = this.activeCell;
+    if (activeCell) {
+      if (this._history) {
+        const lastHistory = this._history.back(activeCell);
+        this._history.updateEditor(activeCell, lastHistory);
```
Comment (krassowski) [inline packages/notebook/src/widget.ts:1293]: I would suggest implementing this as async rather than passing a promise.

```suggestion
        const lastHistory = await this._history.back(activeCell);
        this._history.updateEditor(activeCell, lastHistory);
```

File: packages/notebook/src/actions.tsx [hunk 7]
```diff
@@ -2076,6 +2076,24 @@ export namespace NotebookActions {
       }
     });
   }
+
+  /**
+   * Access last notebook history.
+   *
+   * @param notebook - The target notebook widget.
+   */
+  export function accessLastHistory(notebook: Notebook): any {
```
Comment (krassowski) [inline packages/notebook/src/actions.tsx:2085]:
```suggestion
  export function accessLastHistory(notebook: Notebook): void {
```

same below. But I really it feels like this should return `Promise<void>` - also see comments on implementation in `Notebook` class.

**Review by krassowski — COMMENTED** [review]
I left some early comments on API and naming. Since this is a new feature we will need tests and documentation.

**Comment by andrewfulton9** [comment]
Thanks for the review @krassowski! I'll work on addressing all these comments and update ASAP

File: packages/notebook/src/widget.ts (hunk 6 above)
Comment (andrewfulton9) [reply packages/notebook/src/widget.ts:1293]: Does it make sense to not have them be promises or async at all. As far as I can tell all the calls within the function are synchronous so maybe it makes sense to just return a string. with no await.

File: packages/notebook/src/actions.tsx (hunk 7 above)
Comment (andrewfulton9) [reply packages/notebook/src/actions.tsx]: I'm not sure I understand. What's the advantage of making these Promises?

Comment (krassowski) [reply packages/notebook/src/actions.tsx]: The advantage is that when a third-party code calls `NotebookActions.accessPreviousHistory()` (or respective command) they will be able to await it and only then perform another action. This is useful for downstream automation and performance testing. If the promise is dangling at this level we have no way to know when the action is done. Returning the promise rather than loosing it also makes it easier to unit test it.

Comment (andrewfulton9) [reply packages/notebook/src/actions.tsx]: I probably should have been more clear this question I think depends on the above question that I asked in a different thread:

> Does it make sense to not have them be promises or async at all? As far as I can tell all the calls within the function are synchronous so maybe it makes sense to just return a string. with no await.

If I make the underlying back and forward calls regular synchronous functions, then there shouldn't be any dangling promise, right?

Comment (krassowski) [reply packages/notebook/src/actions.tsx]: Yes, but should they really by synchronous? On on hand requesting history is always async, on the other I see that the current implementation just requests it after each cell execution. This may be sub-optimal (because we generate overhead and wire traffic on each cell execution) and if the response is delayed may lead to subpar UX (being presented with outdated history). My thinking is that it should be async all along. This is, we would invalidate history on every execution but only request it when user asks for it. Then navigating within history can be from cache to improve response time.

Comment (andrewfulton9) [reply packages/notebook/src/actions.tsx]: Oh I see. That makes a lot of sense. I made a push that should address that. Let me know what you think. I am still working on adding tests and docs though.

**Comment by andrewfulton9** [comment]
@krassowski, I've added a test for this. Let me know if you think it's sufficient, or I should add something.

Do you have a recommendation for where to add a mention of this feature for the user documentation? I was thinking docs/source/user/notebook.rst  or docs/source/user/documents_kernels.rst. I'm also seeing that a lot of actions have youtube videos along with the description. Is there any documentation on how to add a video or is that something that has to be done by someone with rights to the Jupyter youtube channel?

**Comment by krassowski** [comment]
If I remember correctly we wanted to move away from YouTube videos for user-facing documentation (https://github.com/jupyterlab/jupyterlab/issues/8776 and https://github.com/jupyterlab/jupyterlab/issues/6823) and are not adding new ones. Either notebook or kernels section of user guide sounds good to me; a single paragraph would do fine, but if you think a video is useful, it would ideally be generated using playwright. This is however a bit difficult because there is no API to start/stop recording at specific time in the test case.

File: packages/notebook/src/history.ts [hunk 8]
```diff
@@ -0,0 +1,365 @@
...
+
+  /**
+   * handle a cell execution
+   */
+  private async _onExecuted(
+    _: unknown,
+    args: { notebook: StaticNotebook; cell: Cell }
+  ): Promise<void> {
+    const cell = args['cell'];
+    await this._retrieveHistory(cell).catch();
```
Comment (krassowski) [inline packages/notebook/src/history.ts:292]: This is still fetching 500 entries of history on each cell execution. I just tried this locally and for my current history it is extra 180kB on the websocket wire for each cell; execute a notebook with 100 cells and you got 18MB noise on websocket. This is compared to execution and reply messages being barely 300-600 bytes. 

I would suggest we only fetch history once when it is first requested by the user.

File: docs/source/user/notebook.rst [hunk 9]
```diff
@@ -151,6 +151,10 @@ and select “New Console for Notebook”:
      <iframe src="https://www.youtube-nocookie.com/embed/eq1l7DBngQQ?rel=0&amp;showinfo=0" frameborder="0" allow="autoplay; encrypted-media" allowfullscreen></iframe>
   </div>
 
+.. _kernel_history:
+
+You can iterate through the kernel history in a document cell using ``Alt Up-Arrow`` and ``Alt Down-Arrow``.
```
Comment (krassowski) [inline docs/source/user/notebook.rst:156]: Can this feature be disabled? Should it be opt-in? I am specifically concerned about sending a lot of history which is possibly originating from root_dir other than current; some changes I suggest in other comments (only requesting on first user action triggering `notebook:access-previous-history-entry`/`notebook:access-next-history-entry`, reducing _n_) would alleviate the issue but it still feels not a great thing to do without warning users about it.

**Comment by krassowski** [comment]
Kicking CI after bot update.

File: packages/notebook/src/history.ts (hunk 8 above)
Comment (andrewfulton9) [reply packages/notebook/src/history.ts:292]: I have it fetching history on each reset so that newly run cells are included in the history without manually appending them. Also, fetching it each time allows a user to access the history from different kernels as well, though they would have to scroll through the history of the current kernel first. What if I implemented something where it only requested a few at a time, requesting more as the user scrolls farther back through the history.

I'm also now wondering if accessing different kernel histories could be considered a security risk with multiple users on jupyter_server? Would users be able to access the kernel history for other users as it is now?

File: packages/notebook/src/history.ts [hunk 10]
```diff
@@ -0,0 +1,365 @@
...
+  }
+
+  /**
+   * retrieve the history from the kernel
+   *
+   * @param cell - The string to use when filtering the data.
+   */
+  private async _retrieveHistory(cell?: Cell): Promise<void> {
+    return await this._kernel
+      ?.requestHistory(Private.initialRequest)
```
[the remark this answers is not in the document]
Comment (andrewfulton9) [reply packages/notebook/src/history.ts:302]: Any suggestion on how this should be handled? It seems easy enough to catch this, but I'm wondering how it should be handled after the catch. Would a notification.warning with autoClose be sufficient?

File: docs/source/user/notebook.rst (hunk 9 above)
Comment (andrewfulton9) [reply docs/source/user/notebook.rst:156]: Yeah, I think that could be a good idea. Could you point me to an example of another feature that is opt-in that I could use as an example?

File: packages/notebook/src/history.ts (hunk 8 above)
Comment (krassowski) [reply packages/notebook/src/history.ts:292]: It is fine to fetch say 10-20 items at a time and then maybe more if there is no overlap with what was fetched before. All I am arguing is that it should never happen on cell execution, but only when user does invoke `notebook:access-previous-history-entry`/`notebook:access-next-history-entry`. We could use cell execution as a heuristic for how many items to fetch; in pseudo-code:

```ts
class {
  onCellExecuted() {
     this._minToFetch += 1;
  }
  
  previousEntry() {
    this._ensureHistorySynced();
    this._setPrevious()
  }
  
  private _ensureHistorySynced() {
    let historyReconcliliated = false;
    let initialToFetch = this._minToFetch + 5;
    let historyChunk = this._fetch(initialToFetch);
    const historyChunks = [];
    while (!historyReconcliliated) {
      if (historyChunk overlaps this._historyCache) {
        historyChunks.push(historyChunk[overlapStart:]);
        historyReconcliliated = true;
      } else {
        // this assumes no execution in the meantime
        historyChunks.push(historyChunk);
        historyChunk = this._fetch(10);
      }
    }
    this._historyCache.push(...reversed(historyChunks));
    this._minToFetch = 0;
  }
  private _minToFetch = 0;
}
```

Though this will be tricky to get right. Do we really want to have history from concurrently executing kernels though?

Maybe using `session`, `start` and `stop` arguments of, `history_request` could help? In particular if we restrict ourselves to single `session`, the problem of cross-session history leakage would be eliminated (but then the history would be more limited).

File: packages/notebook/src/history.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (krassowski) [reply packages/notebook/src/history.ts:302]: I would say that this is even fine to have as `console.warn` - this is a bit pathological case of kernel dying; it is just a bad practice to have uncaught exceptions because these sometimes cause trouble to users with extensions using catch-all handlers (and make user bug reports noisy).

File: docs/source/user/notebook.rst (hunk 9 above)
Comment (krassowski) [reply docs/source/user/notebook.rst:156]: In notebook context, side-by-side rendering, in completer there is documentation panel (though to be honest these are not very well documented).

File: packages/notebook/src/history.ts (hunk 8 above)
Comment (andrewfulton9) [reply packages/notebook/src/history.ts:292]: Yeah I was on the fence on whether or not to include multiple kernel histories or just the history of the notebook kernel, I ultimately decided to include them all since that's how the console history works. I did mostly implement the single notebook kernel history though, and that's actually why cell history is collected on execution. I couldn't figure out another way to get the kernel id number without collecting from the first history item immediately after a cell in the notebook was run, though if we go that way it should probably only call history after the first execution. So it seems to me the options are:

1) include whole history, and only collect on access commands or
2) include only history from current kernel, getting kernel number from first execution of the session

and then with either of those options it can either:

a) grab history in intervals or
b) grab n number (what we currently have)

What are your thoughts?

Comment (andrewfulton9) [reply packages/notebook/src/history.ts]: @rsaim, do you have an opinion on this?

**Comment by andrewfulton9** [comment]
@krassowski, I added in the batch logic for pulling history to this PR. It still includes the whole ipython history. I also made it so that scrolling through the history appears linear, even if history items are added above the cursor. There is an efficiency tradeoff of doing it like this, in that while accessing near history items it's considerably cheaper for network traffic, scrolling deep will also be considerable more expensive.

File: packages/notebook/src/widget.ts [hunk 11]
```diff
@@ -1189,6 +1197,7 @@ export namespace StaticNotebook {
      * - 'none': Attach all cells to the viewport
      */
     windowingMode: 'defer' | 'full' | 'none';
+    accessKernelHistory: boolean;
```
Comment (fcollonval) [inline packages/notebook/src/widget.ts:1200]: Could you add a doc string for this interface?

File: packages/notebook/src/history.ts [hunk 12]
```diff
@@ -0,0 +1,419 @@
...
+  export interface IOptions {
+    /**
+     * The client session used by the foreign handler.
+     */
+    sessionContext: ISessionContext;
+
+    /**
+     * The application language translator.
+     */
+    translator: ITranslator;
```
Comment (fcollonval) [inline packages/notebook/src/history.ts:408]: The constructor supports optional translator but not this interface:

```suggestion
    translator?: ITranslator;
```

File: packages/notebook/src/history.ts [hunk 13]
```diff
@@ -0,0 +1,419 @@
...
+    this._cursor = Math.min(this._filtered.length - 1, this._cursor);
+    const content = this._filtered[this._cursor];
+    return content;
+  }
+
+  /**
+   * Get the next item in the console history.
+   *
+   * @param activeCell - The currently selected Cell in the notebook.
+   * @param update - the promise returned from back or forward
```
Comment (fcollonval) [inline packages/notebook/src/history.ts:217]: Let's drop _promise_ as it is the content

```suggestion
   * @param update - the result from back or forward
```

File: packages/notebook/src/history.ts [hunk 14]
```diff
@@ -0,0 +1,419 @@
...
+  /**
+   * Reset the history navigation state, i.e., start a new history session.
+   */
+  reset(): void;
+
+  /**
+   * Get the next item in the console history.
+   *
+   * @param activeCell - The currently selected Cell in the notebook.
+   * @param update - the promise returned from back or forward
```
Comment (fcollonval) [inline packages/notebook/src/history.ts:69]:
```suggestion
   * @param update - the result from back or forward
```

File: packages/notebook/src/widget.ts [hunk 15]
```diff
@@ -989,6 +991,7 @@ export class StaticNotebook extends WindowedList {
   private _idleCallBack: number | null;
   private _mimetype: string;
   private _mimetypeService: IEditorMimeTypeService;
+  readonly _history: INotebookHistory | undefined;
```
Comment (fcollonval) [inline packages/notebook/src/widget.ts:994]: I would recommend making this one protected

```suggestion
  protected readonly _history: INotebookHistory | undefined;
```

**Review by fcollonval — COMMENTED** [review]
Thanks @andrewfulton9  (and thanks @krassowski for the ping in the call).

i have a couple of minor suggestions.

File: packages/notebook/src/widget.ts (hunk 11 above)
Comment (fcollonval) [reply packages/notebook/src/widget.ts:1200]: And it should be optional as it is changing the API for `set notebookConfig`. This probably requires to modify that setter to use the default value if it is not provided.

```suggestion
    accessKernelHistory?: boolean;
```

File: packages/notebook/src/history.ts [hunk 16]
```diff
@@ -0,0 +1,419 @@
...
+ */
+export class NotebookHistory implements INotebookHistory {
+  /**
+   * Construct a new console history object.
+   */
+  constructor(options: NotebookHistory.IOptions) {
+    this.sessionContext = options.sessionContext;
+    this.translator = options.translator || nullTranslator;
+    void this._handleKernel();
+    this.sessionContext.kernelChanged.connect(this._handleKernel, this);
```
[the remark this answers is not in the document]
Comment (andrewfulton9) [reply packages/notebook/src/history.ts:85]: good point, I think you're right about that

File: packages/notebook/src/widget.ts (hunk 15 above)
Comment (krassowski) [reply packages/notebook/src/widget.ts]: We need a way to access history:

```
      ../notebook/src/actions.tsx(2100,20): error TS2445: Property '_history' is protected and only accessible within class 'StaticNotebook' and its subclasses.
```

Maybe it needs to be just `readolny` (public) but then it should probably not have `_` prefix. This is similar to the story with `footer`. Thoughts?

Comment (andrewfulton9) [reply packages/notebook/src/widget.ts]: Right, it used to be private and notebook had its own back/forward methods that called history from the `Notebook` class which were in turn called from `accessPreviousHistory` and `accessNextHistory` in [actions](https://github.com/jupyterlab/jupyterlab/blob/a37354a42b6d63f64dcb0e2a16216897b5919348/packages/notebook/src/actions.tsx#L2092). Now those just directly use `notebook._history`. So it seems we should either revert the code to keep `_history` private or just make it readonly and lose the `_` prefix. I would lean towards the second option.

Comment (krassowski) [reply packages/notebook/src/widget.ts]: > I would lean towards the second option.

Let's do that to unblock this pull request. We will still have some time to change things before 4.1 release if we decide it was not the right call.

File: packages/notebook/src/history.ts [hunk 17]
```diff
@@ -0,0 +1,420 @@
...
+    this._hasSession = false;
+    this._placeholder = '';
+    this.toRequest = this.requestBatchSize;
+  }
+
+  /**
+   * Fetches a subsequent batch of history. Updates the filtered history and cursor to correct place in history,
+   * accounting for potentially new history items above it.
+   */
+  async fetchBatch() {
```
Comment (krassowski) [inline packages/notebook/src/history.ts:254]: private/protected?

Would it be worth adding a jest unit test for this logic (indirectly, by testing `back` which is the public API for it)?

Comment (andrewfulton9) [reply packages/notebook/src/history.ts]: Good call. I changed it to private, and added a test as well

**Comment by krassowski** [comment]
bot please update snapshots

File: packages/notebook/test/history.spec.ts [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+        ipySessionContext.shutdown()
+      ]);
+    });
+
+    describe('#history', () => {
+      it('should emit when Markdown and code cells are run', async () => {
+        for (let i = 0; i < 12; i++) {
+          let cell = widget.widgets[i] as CodeCell;
+          let source = 'print(this is input ${i})';
+          cell.model.sharedModel.setSource(source);
```
Comment (krassowski) [inline packages/notebook/test/history.spec.ts:86]: This test is failing here, it looks like the cells need to be explicitly added to the notebook.

```
@jupyterlab/notebook › NotebookActions › #history › should emit when Markdown and code cells are run

TypeError: Cannot read properties of undefined (reading 'model')

    at Object.<anonymous> (test/history.spec.ts:86:16)
````

File: galata/test/documentation/general.test.ts-snapshots/jupyterlab-documentation-linux.png
Comment (krassowski) [inline galata/test/documentation/general.test.ts-snapshots/jupyterlab-documentation-linux.png:1]: This one appears unrelated and can be reverted to reduce noise.

File: packages/notebook/src/widget.ts (hunk 1 above)
Comment (andrewfulton9) [reply packages/notebook/src/widget.ts:1049]: Maybe `kernelHistory` or `executedInputHistory`?

Comment (krassowski) [reply packages/notebook/src/widget.ts:1049]: `kernelHistory` sounds good to me

File: packages/notebook/src/history.ts [hunk 19]
```diff
@@ -0,0 +1,420 @@
...
+
+  /**
+   * Translator to be used for warnings
+   */
+  readonly translator: ITranslator;
+
+  /**
+   * The number of history items to request.
+   */
+  private toRequest: number;
```
Comment (krassowski) [inline packages/notebook/src/history.ts:103]: Nitpick: elsewhere we use underscore prefix for private properties (though not strictly required).

File: packages/notebook/src/history.ts [hunk 20]
```diff
@@ -0,0 +1,420 @@
...
+
+  /**
+   * Kernel session number for filtering
+   */
+  get kernelSession(): string {
+    return this._kernelSession;
+  }
+
+  /**
+   * Get whether the console history manager is disposed.
```
Comment (krassowski) [inline packages/notebook/src/history.ts:149]:
```suggestion
   * Get whether the notebook history manager is disposed.
```

File: packages/notebook/src/history.ts [hunk 21]
```diff
@@ -0,0 +1,420 @@
...
+   */
+  get isDisposed(): boolean {
+    return this._isDisposed;
+  }
+
+  /**
+   * Dispose of the resources held by the console history manager.
```
Comment (krassowski) [inline packages/notebook/src/history.ts:156]:
```suggestion
   * Dispose of the resources held by the notebook history manager.
```

File: packages/notebook/src/history.ts [hunk 22]
```diff
@@ -0,0 +1,420 @@
...
+      this.editor = activeCell.editor;
+      this._placeholder = this._editor?.model.sharedModel.getSource() || '';
+      // Filter the history with the placeholder string.
+      this.setFilter(this._placeholder);
+      this._cursor = this._filtered.length - 1;
+    }
+  }
+
+  /**
+   * Get the previous item in the console history.
```
Comment (krassowski) [inline packages/notebook/src/history.ts:182]:
```suggestion
   * Get the previous item in the notebook history.
```

File: packages/notebook/src/history.ts [hunk 23]
```diff
@@ -0,0 +1,420 @@
...
+   *
+   * @param activeCell - The currently selected Cell in the notebook.
+   *
+   * @returns A Promise for console command text or `undefined` if unavailable.
```
Comment (krassowski) [inline packages/notebook/src/history.ts:186]:
```suggestion
   * @returns A Promise resolving to the historical cell content text.
```

File: packages/notebook/src/history.ts [hunk 24]
```diff
@@ -0,0 +1,420 @@
...
+    if (this._cursor < 0) {
+      await this.fetchBatch();
+    }
+    this._cursor = Math.max(0, this._cursor);
+    const content = this._filtered[this._cursor];
+    return content;
+  }
+
+  /**
+   * Get the next item in the console history.
```
Comment (krassowski) [inline packages/notebook/src/history.ts:200]:
```suggestion
   * Get the next item in the notebook history.
```

File: packages/notebook/src/history.ts [hunk 25]
```diff
@@ -0,0 +1,420 @@
...
+   *
+   * @param activeCell - The currently selected Cell in the notebook.
+   *
+   * @returns A Promise for console command text or `undefined` if unavailable.
```
Comment (krassowski) [inline packages/notebook/src/history.ts:204]:
```suggestion
   * @returns A Promise resolving to the historical cell content text.
```

File: packages/notebook/src/history.ts [hunk 26]
```diff
@@ -0,0 +1,420 @@
...
+   */
+  async back(activeCell: Cell): Promise<string> {
+    await this.checkSession(activeCell);
+    --this._cursor;
+    if (this._cursor < 0) {
+      await this.fetchBatch();
+    }
+    this._cursor = Math.max(0, this._cursor);
+    const content = this._filtered[this._cursor];
+    return content;
```
Comment (krassowski) [inline packages/notebook/src/history.ts:196]: The typing suggests that it will never return undefined as it will stop at the oldest entry. But what happens if there is no history yet? In a very fresh installation, `this._filtered[0]` would resolve to `undefined`, right? Will the downstream error out due to incorrect typing? I would suggest updating the return type to `Promise<string | undefined>` and making any changes that may be needed in the calling functions accordingly.

File: packages/notebook/src/history.ts [hunk 27]
```diff
@@ -0,0 +1,420 @@
...
+   * @param activeCell - The currently selected Cell in the notebook.
+   *
+   * @returns A Promise for console command text or `undefined` if unavailable.
+   */
+  async forward(activeCell: Cell): Promise<string> {
+    await this.checkSession(activeCell);
+    ++this._cursor;
+    this._cursor = Math.min(this._filtered.length - 1, this._cursor);
+    const content = this._filtered[this._cursor];
+    return content;
```
Comment (krassowski) [inline packages/notebook/src/history.ts:211]: Again, what happens if history is empty?

File: packages/notebook/src/history.ts [hunk 28]
```diff
@@ -0,0 +1,420 @@
...
+  }
+
+  /**
+   * Get the next item in the console history.
```
Comment (krassowski) [inline packages/notebook/src/history.ts:215]:
```suggestion
   * Update the editor of the cell with provided text content.
```

File: packages/notebook/src/history.ts [hunk 29]
```diff
@@ -0,0 +1,420 @@
...
+   *
+   * @param activeCell - The currently selected Cell in the notebook.
+   * @param update - the result from back or forward
```
Comment (krassowski) [inline packages/notebook/src/history.ts:218]: While `update` is technically correct, it would be more intuitive for me if it was named `content`

File: docs/source/user/notebook.rst [hunk 30]
```diff
@@ -151,6 +151,10 @@ and select “New Console for Notebook”:
      <iframe src="https://www.youtube-nocookie.com/embed/eq1l7DBngQQ?rel=0&amp;showinfo=0" frameborder="0" allow="autoplay; encrypted-media" allowfullscreen></iframe>
   </div>
 
+.. _kernel_history:
+
+You can iterate through the kernel history in a document cell using ``Alt Up-Arrow`` and ``Alt Down-Arrow``. To use this feature, enable it in the notebook settings.
```
Comment (krassowski) [inline docs/source/user/notebook.rst:156]:
```suggestion
You can iterate through the kernel history in a document cell using ``Alt Up-Arrow`` and ``Alt Down-Arrow``. To use this feature, enable kernel history access in the notebook settings.
```

File: packages/notebook-extension/schema/tracker.json [hunk 31]
```diff
@@ -801,6 +811,12 @@
       "description": "'defer': Improve loading time - Wait for idle CPU cycles to attach out of viewport cells - 'full': Best performance with side effects - Attach to the DOM only cells in viewport - 'none': Worst performance without side effects - Attach all cells to the viewport",
       "enum": ["defer", "full", "none"],
       "default": "defer"
+    },
+    "accessKernelHistory": {
+      "title": "Enable kernel history access.",
```
Comment (krassowski) [inline packages/notebook-extension/schema/tracker.json:816]: Simpler?

```suggestion
      "title": "Kernel history access",
```

**Review by krassowski — COMMENTED** [review]
This is almost there, just minor changes, mostly updating documentation strings + one issue with translation and narrowing the public API.

File: packages/notebook/src/history.ts (hunk 26 above)
Comment (andrewfulton9) [reply packages/notebook/src/history.ts:196]: The source of the active cell is added to `this._filtered` in `setFilter` and it isn't possible to access `_filtered` before `setFilter` is run so it shouldn't be possible. I can update it just in case though.

Comment (krassowski) [reply packages/notebook/src/history.ts:190]: In that case no need to change code, but adding a comment would be helpful.

Comment (krassowski) [reply packages/notebook/src/history.ts:190]: Or ideally a unit test case

File: packages/notebook/src/widget.ts [hunk 32]
```diff
@@ -234,6 +235,7 @@ export class StaticNotebook extends WindowedList {
     this._updateNotebookConfig();
     this._mimetypeService = options.mimeTypeService;
     this.renderingLayout = options.notebookConfig?.renderingLayout;
+    this.kernelHistory = options.kernelHistory;
```
Comment (krassowski) [inline packages/notebook/src/widget.ts:238]: Currently this PR the integration test fails because the factory is still passing `history` rather than `kernelHistory`, see the other two comments.

**Review by krassowski — APPROVED** [review]
Thank you @andrewfulton9!
